### PR TITLE
Prevent Brush from being repeatedly allocated

### PR DESCRIPTION
**Pull request by Synchronicity89**

### CODE REVIEW

**Review by Synchronicity89 — COMMENTED** [review]
I am changing this back to draft PR to address these issues with it:

1. User might be able to change the color of neurons and synapses down the road, so want to avoid hardcoding of colors like Blue, Red
2. The colors could be added to XML where possible and only allocated once a Module is instantiated.

**Comment by FutureAIGuru** [comment]
Hi Kevin—

The color of neurons and synapses  is a rainbow based on the value as in NeuronArrayView Line 490.
If you want to call, that might be more efficient.

Charles J. Simon, BSEE MSCs
Founder
[Text  Description automatically generated with medium confidence][Logo  Description automatically generated][Text, application  Description automatically generated]
455 Massachusetts Avenue, Suite 120
Washington, D.C. 20001
425-765-8162
https://FutureAI.guru<https://futureai.guru/>

WASHINGTON, D.C. – SPOKANE, WA.

This e-mail is for the sole use of the intended recipient(s).  It contains information that may be confidential and/or legally privileged.  If you believe that it has been sent in error, please notify the sender immediately by reply e-mail and destroy all copies of this message.  Any disclosure, copying, distribution, or use of this information by anyone other than the intended recipient is prohibited.

**Comment by Synchronicity89** [comment]
Ah, I will look at NeuronView and its function GetNeuronColor.
I'll get metrics on any performance improvement I get.  I'll also consider users may ask for some kind of flexibility as to colors (for example for color blind people), as well as allocating brushes only as needed.

**Comment by Synchronicity89** [comment]
Updated the commit with latest changes, but have not got metrics on performance yet to prove that it helps.  I am getting big lags on a system with 32 GB RAM and an i7. Need to make sure that my changes aren't causing them.  Will test on a i9 system with 64GB RAM.

**Comment by FutureAIGuru** [comment]
Thanks for the update. UI performance should be fine unless the network is huge. We use 16M 4-core ryzens and performance is fine. That doesn’t mean that your code is the issue. The last time I fetched your stuff, it ran fine too.

File: BrainSimulator/Modules/ModuleMove.cs [hunk 1]
```diff
@@ -51,7 +51,7 @@ public override void Fire()
                 moved = m2D.Move(motionX, motionY);
 
             Module2DModel m2DModel = (Module2DModel)FindModleu(typeof(Module2DModel));
-            if (m2DModel != null && moved && motionX != 0 || motionY != 0)
+            if (m2DModel != null && moved && (motionX != 0 || motionY != 0))
```
Comment (NickFutureAI) [inline BrainSimulator/Modules/ModuleMove.cs:54]: Good job catching this logical error

**Comment by NickFutureAI** [comment]
Thank you for this work but in its current state, I cannot approve this request until the below-mentioned "Bugs Created" are corrected.

Features:
Color Cache
- Memory Improvement. About 16 fewer MBs were used in BasicNeurons, same with Hebbian Synapses. CPU usage was comparable once fully loaded.

Functionally equivalent besides:
- Change makes neurons out of use same color as in use. This makes it more difficult to determine what is and is not related to the module.
- The opacity setting is likely being missed or now broken. Opacity is set through the NeuronView

Bugs Created:
- Change makes neurons out of use same color as in use (see above for more details)
- The class name of Cache creates a naming conflict with System.Net.Cache

Bugs Corrected:
- Caught logical error in ModuleMove line 54

**Review by Synchronicity89 — COMMENTED** [review]
This change probably doesn't belong in this PR.  Not sure how to remove it, but the changes probably could be removed after PR is merged, if they are not wanted.

**Comment by Synchronicity89** [comment]
Converted it back to draft.  I got an exception when I accidently dragged the mouse.  Freezing graphics elements is an interesting idea and it has already been used once in this project to Freeze an image.  Introducing Freezing for other graphics objects is probably an architectural decision, and would have to be made robust.  One way to implement it might be to create methods like this SolidColorBrush SetOpacity(SolidColorBrush brush, double opacity) { .... }

This would return a brush that was either the original brush, or if it was frozen, a clone, with the new opacity. 

It would be nice if some brushes with different opacities were also cached.  The cache would have to be redesigned to do so.

I will experiment in my fork with a less risky change.   I will try XAML markup to introduce Freezing and see what that does.

BTW: My next PR might be this feature:  Importing ONNX model data and generating synaptic weights and neuron charges from that.  I have made a POC that shows I can access the raw ONNX data and generate a network from it (thus far in an arbitrary way).  So far I have generated this attached XML.
[FirstMostlyGeneratedFromONNX.zip](https://github.com/FutureAIGuru/BrainSimII/files/8119207/FirstMostlyGeneratedFromONNX.zip)

The ONNX import POC will be soon be available in the stressTest branch of my fork.  It will stress test the Brush Cache by having thousands of colors based off of ONNX model data.  My vision/mission with ONNX import would be to bring in ONNX models into BrainSimII and see them in action.

**Comment by Synchronicity89** [comment]
I am finished working on this PR except for fixing any bugs or implementing requested changes.  

It does what I wanted it to do.   It caches frozen brushes of each opacity and color used.  It also Freezes all Freezable to speed up graphics.  I realize that is an architectural type of change, so it should be used only if that architectural change is opted for.   

If nothing else its a proof of concept that caching and freezing Freezable objects appears to work and frees up memory.  If I run into a crash due to an invalid operation on a frozen object I'll mark this PR draft again.  But so far it runs great.

What I may work on next: In my stressTest branch I might continue to work on importing ONNX files, possibly using Microsoft's Microsoft.ML.OnnxRuntime  nuget package to retrain the ONNX-imported spiking neural network.  As of this moment, in my fork's stressTest branch the imported ONNX weights are used in an arbitrary manner just to prove they can be retrieved from an onnx file. So it is early days for this attempt.

Caveat: Some gitignore changes in this PR may have to be undone or left out.

**Comment by NickFutureAI** [comment]
This looks great, all requested fixes were made. I am passing this off to Charles @FutureAIGuru for final approval.

**Pull request closed without merging**